### DISCUSSION TITLE
output is in correct

### DISCUSSION THREAD

**Discussion by Nehad84** [rfennell/AzurePipelines#1323] [Q&A] [answered]
Hi,
I'm using this path: $(System.DefaultWorkingDirectory)/releasenotes.md in both generate release note in the output file field,  and in wiki updater in Source File with Data in file selected, it fails <##[error]Cannot find the file </home/vsts/work/r1/a/releasenotes.md> and when I use it as variabled passed from the generate release note task, it write down what I put in the variable field, even I used the reference

**Comment by rfennell** [comment]
Can you upload the build logs so I can have a detailed look?

**Comment by Nehad84** [comment]
Thank you so much for your response, I think teh file is not created because the previous step can't parse handelbars, here is the build yaml `steps:
- task: richardfennellBM.BM-VSTS-XplatGenerateReleaseNotes.XplatGenerate-Release-Notes.XplatGenerateReleaseNotes@3
  displayName: 'Generate Release Notes based on Release Comparison API'
  inputs:
    outputfile: '$(Build.ArtifactStagingDirectory)/releasenotes.md'
    templateLocation: InLine
    inlinetemplate: |
     ## Build {{buildDetails.buildNumber}}
     * **Branch**: {{buildDetails.sourceBranch}}
     * **Tags**: {{buildDetails.tags}}
     * **Completed**: {{buildDetails.finishTime}}
     * **Previous Build**: {{compareBuildDetails.buildNumber}}
     
     
     # Global list of WI with PRs, parents and children
     {{#forEach this.workItems}}
     {{#if isFirst}}### WorkItems {{/if}}
     *  **{{this.id}}**  {{lookup this.fields 'System.Title'}}
        - **WIT** {{lookup this.fields 'System.WorkItemType'}}
        - **Tags** {{lookup this.fields 'System.Tags'}}
        - **Assigned** {{#with (lookup this.fields 'System.AssignedTo')}} {{displayName}} {{/with}}
        - **Description** {{{lookup this.fields 'System.Description'}}}
        - **PRs**
     {{#forEach this.relations}}
     {{#if (contains this.attributes.name 'Pull Request')}}
     {{#with (lookup_a_pullrequest ../../pullRequests  this.url)}}
           - {{this.pullRequestId}} - {{this.title}}
     {{/with}}
     {{/if}}
     {{/forEach}}
        - **Parents**
     {{#forEach this.relations}}
     {{#if (contains this.attributes.name 'Parent')}}
     {{#with (lookup_a_work_item ../../relatedWorkItems  this.url)}}
           - {{this.id}} - {{lookup this.fields 'System.Title'}}
           {{#forEach this.relations}}
           {{#if (contains this.attributes.name 'Parent')}}
           {{#with (lookup_a_work_item ../../../../relatedWorkItems  this.url)}}
              - {{this.id}} - {{lookup this.fields 'System.Title'}}
           {{/with}}
           {{/if}}
           {{/forEach}}
     {{/with}}
     {{/if}}
     {{/forEach}}
        - **Children**
     {{#forEach this.relations}}
     {{#if (contains this.attributes.name 'Child')}}
     {{#with (lookup_a_work_item ../../relatedWorkItems  this.url)}}
           - {{this.id}} - {{lookup this.fields 'System.Title'}}
     {{/with}}
     {{/if}}
     {{/forEach}}
       
    stopOnRedeploy: false
    customHandlebarsExtensionCode: |
     const handlebars = require("handlebars");
     module.exports = {
         count_workitems_by_type: function (array, typeName) {
             return array.filter(wi => wi.fields['System.WorkItemType'] === typeName).length;
         },
         replace_text: function (msg, match, replacement) {
             return msg.replace(match, replacement);
         }
     };
    customHandlebarsExtensionFolder: 'https://github.com/rfennell/AzurePipelines/tree/main/SampleTemplates/XplatGenerateReleaseNotes%20(Node%20based)/Version%203'
    maxRetries: 100
    stopOnError: true
`

and here's the output:
`2022-08-19T20:25:17.1906393Z - PR 655 does not have a lastMergeCommit
2022-08-19T20:25:17.1906636Z Enriching known Pull Requests
2022-08-19T20:29:15.3890113Z Total Builds: [1]
2022-08-19T20:29:15.3890591Z Total Commits: [420]
2022-08-19T20:29:15.3890998Z Total Workitems: [100]
2022-08-19T20:29:15.3891427Z Total Related Workitems (Parent/Children): [0]
2022-08-19T20:29:15.3891847Z Total Release Tests: [0]
2022-08-19T20:29:15.3892242Z Total Tests: [0]
2022-08-19T20:29:15.3892626Z Total Manual Test Runs: [0]
2022-08-19T20:29:15.3893069Z Total Manual Test Configurations: [0]
2022-08-19T20:29:15.3893491Z Total Pull Requests: [102]
2022-08-19T20:29:15.3893895Z Total Indirect Pull Requests: [0]
2022-08-19T20:29:15.3894317Z Total Associated Test WI: [55]
2022-08-19T20:29:15.3894720Z Total Consumed Artifacts: [0]
2022-08-19T20:29:15.3895126Z Total WIQL Workitems: [0]
2022-08-19T20:29:15.9516322Z Generating the release notes, the are 1 template(s) to process
2022-08-19T20:29:15.9523830Z Using in-line template
2022-08-19T20:29:15.9760453Z Loading handlebars-helpers extension
2022-08-19T20:29:16.0443414Z Writing output file /home/vsts/work/_tasks/XplatGenerateReleaseNotes_7b0a5887-75c4-4ade-a915-f92a93c934cb/3.85.2/customHandlebarsExtension.js
2022-08-19T20:29:16.0444525Z Finished writing output file /home/vsts/work/_tasks/XplatGenerateReleaseNotes_7b0a5887-75c4-4ade-a915-f92a93c934cb/3.85.2/customHandlebarsExtension.js
2022-08-19T20:29:16.0446618Z Loaded handlebars extension file
2022-08-19T20:29:16.0559169Z ##[error]Error Processing handlebars [Error: Parse error on line 37:
...{{/if}}{{/forEach}}
----------------------^
Expecting 'OPEN_INVERSE_CHAIN', 'INVERSE', 'OPEN_ENDBLOCK', got 'EOF']
2022-08-19T20:29:16.0569801Z ##[error]Error Processing handlebars [Error: Parse error on line 37:
...{{/if}}{{/forEach}}
----------------------^
Expecting 'OPEN_INVERSE_CHAIN', 'INVERSE', 'OPEN_ENDBLOCK', got 'EOF']
2022-08-19T20:29:16.0790131Z ##[error]Error Processing handlebars [Error: Parse error on line 37:
...{{/if}}{{/forEach}}
----------------------^
Expecting 'OPEN_INVERSE_CHAIN', 'INVERSE', 'OPEN_ENDBLOCK', got 'EOF']
2022-08-19T20:29:16.0855238Z ##[section]Finishing: Generate Release Notes based on Release Comparison API`

**Reply by rfennell** [reply]
Looks like you have a custom handlbars extension defined that has a error in it, and this is killing the release notes task so no output file is being generated

> 2022-08-19T20:29:16.0446618Z Loaded handlebars extension file 
> 2022-08-19T20:29:16.0559169Z ##[error]Error Processing handlebars [Error: Parse error on line 37: ...{{/if}}{{/forEach}} ----------------------^ Expecting 'OPEN_INVERSE_CHAIN', 'INVERSE', 'OPEN_ENDBLOCK', got 'EOF'] 2022-08-19T20:29:16.0569801Z ##[error]Error Processing handlebars [Error: Parse error on line 37: ...{{/if}}{{/forEach}} ----------------------^ Expecting 'OPEN_INVERSE_CHAIN', 'INVERSE', 'OPEN_ENDBLOCK', got 'EOF'] 
> 2022-08-19T20:29:16.0790131Z ##[error]Error Processing handlebars [Error: Parse error on line 37: ...{{/if}}{{/forEach}} ----------------------^ Expecting 'OPEN_INVERSE_CHAIN', 'INVERSE', 'OPEN_ENDBLOCK', got 'EOF'] 

You need to fix the extension, or remove it.

You might find [the local tester](https://github.com/rfennell/AzurePipelines/tree/main/Extensions/XplatGenerateReleaseNotes/V3/testconsole) a help in debugging this

**Comment by Nehad84** [comment, marked as answer]
it doesn't work with or without the extension

**Reply by rfennell** [reply]
I will test your template on my test system

**Reply by Nehad84** [reply]
I used the exact template from here:

Extensions/XplatGenerateReleaseNotes/readme.md

Nothing more nothing less

**Reply by rfennell** [reply]
It as a quick fix, you had not closed one of your foreach loops in your template. 

I found thisby running your template, and custom extension, in [my local tester](https://github.com/rfennell/AzurePipelines/tree/main/Extensions/XplatGenerateReleaseNotes/V3/testconsole) and got the same error. A quick scan of the template saw the mismatch. 

The local tester is really useful for this type of debug, it is why I wrote it. I suggest you always use it when developing a handlbars template a or custom extension.

The working template is

```
     ## Build {{buildDetails.buildNumber}}
     * **Branch**: {{buildDetails.sourceBranch}}
     * **Tags**: {{buildDetails.tags}}
     * **Completed**: {{buildDetails.finishTime}}
     * **Previous Build**: {{compareBuildDetails.buildNumber}}
     
     
     # Global list of WI with PRs, parents and children
     {{#forEach this.workItems}}
     {{#if isFirst}}### WorkItems {{/if}}
     *  **{{this.id}}**  {{lookup this.fields 'System.Title'}}
        - **WIT** {{lookup this.fields 'System.WorkItemType'}}
        - **Tags** {{lookup this.fields 'System.Tags'}}
        - **Assigned** {{#with (lookup this.fields 'System.AssignedTo')}} {{displayName}} {{/with}}
        - **Description** {{{lookup this.fields 'System.Description'}}}
        - **PRs**
     {{#forEach this.relations}}
     {{#if (contains this.attributes.name 'Pull Request')}}
     {{#with (lookup_a_pullrequest ../../pullRequests  this.url)}}
           - {{this.pullRequestId}} - {{this.title}}
     {{/with}}
     {{/if}}
     {{/forEach}}
        - **Parents**
     {{#forEach this.relations}}
     {{#if (contains this.attributes.name 'Parent')}}
     {{#with (lookup_a_work_item ../../relatedWorkItems  this.url)}}
           - {{this.id}} - {{lookup this.fields 'System.Title'}}
           {{#forEach this.relations}}
           {{#if (contains this.attributes.name 'Parent')}}
           {{#with (lookup_a_work_item ../../../../relatedWorkItems  this.url)}}
              - {{this.id}} - {{lookup this.fields 'System.Title'}}
           {{/with}}
           {{/if}}
           {{/forEach}}
     {{/with}}
     {{/if}}
     {{/forEach}}
        - **Children**
     {{#forEach this.relations}}
     {{#if (contains this.attributes.name 'Child')}}
     {{#with (lookup_a_work_item ../../relatedWorkItems  this.url)}}
           - {{this.id}} - {{lookup this.fields 'System.Title'}}
     {{/with}}
     {{/if}}
     {{/forEach}}
     {{/forEach}}
   ```

**Reply by Nehad84** [reply]
thank you so much for your help, it works now

**Comment by Nehad84** [comment]
Hi @rfennell, So sorry to bother you again, I'm using the below template it's as same as above with some modification - trying to add summary before the details - , but when I run the local test, I got this error line 64 is line 32 in the old template - also even when I run the only the first 5 lines of the template it took more than 3 hours to fail

`##vso[task.issue type=error;]Error Processing handlebars [Error: Parse error on line 64:%0A...tributes.name 'Pull Request')}}{{#with %0A-----------------------^%0AExpecting 'CLOSE_RAW_BLOCK', 'CLOSE', 'CLOSE_UNESCAPED', 'OPEN_SEXPR', 'CLOSE_SEXPR', 'ID', 'EQUALS', 'OPEN_BLOCK_PARAMS', 'STRING', 'NUMBER', 'BOOLEAN', 'UNDEFINED', 'NULL', 'DATA', 'SEP', got 'INVALID']`

Template: 
`## Build {{buildDetails.buildNumber}}
* **Branch**: {{buildDetails.sourceBranch}}
* **Tags**: {{buildDetails.tags}}
* **Completed**: {{buildDetails.finishTime}}
* **Previous Build**: {{compareBuildDetails.buildNumber}}

## Release summary - Effect of the release (List of Released WorkItems)
{{#forEach pullRequests}}
* User Stories
{{#forEach this.workItems}}
{{#if contains this.fields 'System.WorkItemType' 'User story'}}
   *  **{{this.id}}**  {{lookup this.fields 'System.Title'}}
{{/if}}
{{/forEach}}
{{/forEach}}

* Fixed Bugs
{{#forEach this.workItems}}

{{#if contains this.fields 'System.WorkItemType' 'Bug'}}
{{#if contains this.fields 'System.State' 'Closed'}}
   *  **{{this.id}}**  {{lookup this.fields 'System.Title'}}
{{/if}}
{{/if}}
{{/forEach}}


* Known Bugs
{{#forEach this.workItems}}

{{#if contains this.fields 'System.WorkItemType' 'Bug'}}
{{#if contains this.state 'Open'}}
   *  **{{this.id}}**  {{lookup this.fields 'System.Title'}}
{{/if}}
{{/if}}
{{/forEach}}

## Associated Pull Requests ({{pullRequests.length}})
{{#forEach pullRequests}}
* **[{{this.pullRequestId}}]({{replace (replace this.url "_apis/git/repositories" "_git") "pullRequests" "pullRequest"}})** {{this.title}}
* Associated Work Items
{{#forEach this.associatedWorkitems}}
   {{#with (lookup_a_work_item ../../relatedWorkItems this.url)}}
    - [{{this.id}}]({{replace this.url "_apis/wit/workItems" "_workitems/edit"}}) - {{lookup this.fields 'System.Title'}}
   {{/with}}
{{/forEach}}
* Associated Commits (this includes commits on the PR source branch not associated directly with the build)
{{#forEach this.associatedCommits}}
    - [{{this.commitId}}]({{this.remoteUrl}}) -  {{this.comment}}
{{/forEach}}
{{/forEach}}

# Global list of WI with PRs, parents and children
{{#forEach this.workItems}}
{{#if isFirst}}### WorkItems {{/if}}
*  **{{this.id}}**  {{lookup this.fields 'System.Title'}}
   - **WIT** {{lookup this.fields 'System.WorkItemType'}}
   - **Tags** {{lookup this.fields 'System.Tags'}}
   - **Assigned** {{#with (lookup this.fields 'System.AssignedTo')}} {{displayName}} {{/with}}
   - **Description** {{{lookup this.fields 'System.Description'}}}
   - **Aceptance Criteria** {{lookup this.fields 'System.AcceptanceCriteria}}
   - **PRs**
{{#forEach this.relations}}
{{#if (contains this.attributes.name 'Pull Request')}}
{{#with (lookup_a_pullrequest ../../pullRequests  this.url)}}
      - {{this.pullRequestId}} - {{this.title}}
{{/with}}
{{/if}}
{{/forEach}}
   - **Parents**
{{#forEach this.relations}}
{{#if (contains this.attributes.name 'Parent')}}
{{#with (lookup_a_work_item ../../relatedWorkItems  this.url)}}
      - {{this.id}} - {{lookup this.fields 'System.Title'}}
      {{#forEach this.relations}}
      {{#if (contains this.attributes.name 'Parent')}}
      {{#with (lookup_a_work_item ../../../../relatedWorkItems  this.url)}}
         - {{this.id}} - {{lookup this.fields 'System.Title'}}
      {{/with}}
      {{/if}}
      {{/forEach}}
{{/with}}
{{/if}}
{{/forEach}}
   {{!-- - **Children**
{{#forEach this.relations}}
{{#if (contains this.attributes.name 'Child')}}
{{#with (lookup_a_work_item ../../relatedWorkItems  this.url)}}
      - {{this.id}} - {{lookup this.fields 'System.Title'}}
{{/with}}
{{/if}}
{{/forEach}} --}}
   - **Tested By**
{{#forEach this.relations}}
{{#if (contains this.attributes.name 'Tested By')}}
{{#with (lookup_a_work_item ../../testedByWorkItems  this.url)}}
      - {{this.id}} - {{lookup this.fields 'System.Title'}}
{{/with}}
{{/if}}
{{/forEach}}
{{/forEach}}

# Global list of CS ({{commits.length}})
{{#forEach commits}}
{{#if isFirst}}### Associated commits{{/if}}
* ** ID{{this.id}}**
   -  **Message:** {{this.message}}
   -  **Commited by:** {{this.author.displayName}}
   -  **FileCount:** {{this.changes.length}}
{{#forEach this.changes}}
      -  **File path (TFVC or TfsGit):** {{this.item.path}}
      {{!-- -  **File filename (GitHub):** {{this.filename}} --}}
{{/forEach}}
{{/forEach}}

{{!-- ## List of WI returned by WIQL ({{queryWorkItems.length}})
{{#forEach queryWorkItems}}
*  **{{this.id}}** {{lookup this.fields 'System.Title'}}
{{/forEach}}

## Manual Test Plans
| Run ID | Name | State | Total Tests | Passed Tests |
| --- | --- | --- | --- | --- |
{{#forEach manualTests}}
| [{{this.id}}]({{this.webAccessUrl}}) | {{this.name}} | {{this.state}} | {{this.totalTests}} | {{this.passedTests}} |
{{/forEach}}

## Global list of ConsumedArtifacts ({{consumedArtifacts.length}})
| Category | Type | Version Name | Version Id | Commits | Workitems |
|-|-|-|-|-|-|
{{#forEach consumedArtifacts}}
 |{{this.artifactCategory}} | {{this.artifactType}} | {{#if versionName}}{{versionName}}{{/if}} | {{truncate versionId 7}} | {{#if this.commits}} {{this.commits.length}} {{/if}} | {{#if this.workitems}} {{this.workitems.length}} {{/if}} |
{{/forEach}} --}}

`

**Comment by rfennell** [comment]
All you can do is add and remove lines from the template until the error disappears to find the exact line that is causing the problem.

You may have to add a null/empty check protector as some the line so it is only run if a field is populated, but you need to know the failing line first
